kubectl-fdb/cmd/cordon.go

```diff
@@ -116,54 +122,111 @@ kubectl fdb cordon -c cluster --node-selector machine=a,disk=fast
+				operationFailed = true
 				continue
 			}
 
-			processGroup, ok := pod.Labels[cluster.GetProcessGroupIDLabel()]
-			if !ok {
-				fmt.Printf("could not fetch process group ID from Pod: %s\n", pod.Name)
-				continue
+			var processGroups []string
+			for _, pod := range pods.Items {
+				// With the field selector above this shouldn't be required, but it's good to
+				// have a second check.
+				if pod.Spec.NodeName != node {
+					cmd.Printf("Pod: %s is not running on node %s will be ignored\n", pod.Name, node)
+					continue
+				}
+
+				if cluster.ContainsPod(pod) {
+					processGroup, ok := pod.Labels[cluster.GetProcessGroupIDLabel()]
+					if !ok {
+						cmd.Printf("could not fetch process group ID from Pod: %s\n", pod.Name)
+						continue
+					}
+					processGroups = append(processGroups, processGroup)
+				}
+			}
+			err = replaceProcessGroups(kubeClient, cluster.Name, processGroups, namespace, withExclusion, wait, false, true, sleep)
+			if err != nil {
+				operationFailed = true
+				errors = append(errors, fmt.Sprintf("unable to cordon all Pods for cluster %s\n", cluster.Name))
 			}
-			processGroups = append(processGroups, processGroup)
 		}
 	}
+	if operationFailed {
+		return fmt.Errorf("following operation failed, please check and retry: \n, %s", errors)
+	}
+	return nil
+}
 
-	return replaceProcessGroups(kubeClient, cluster.Name, processGroups, namespace, withExclusion, wait, false, true, sleep)
+func fetchPods(kubeClient client.Client, inputClusterName string, namespace string, node string, clusterLabel string) (corev1.PodList, error) {
+	var pods corev1.PodList
+	var err error
+	if inputClusterName == "" {
+		err = kubeClient.List(ctx.Background(), &pods,
+			client.InNamespace(namespace),
+			client.HasLabels([]string{clusterLabel}),
+			client.MatchingFieldsSelector{
+				Selector: fields.OneTermEqualSelector("spec.nodeName", node),
+			})
+	} else {
+		cluster, error := loadCluster(kubeClient, namespace, inputClusterName)
+		if error != nil {
+			return pods, fmt.Errorf("unable to load cluster: %s, skipping", inputClusterName)
+		}
+		err = kubeClient.List(ctx.Background(), &pods,
+			client.InNamespace(namespace),
+			client.MatchingLabels(cluster.GetMatchLabels()),
+			client.MatchingFieldsSelector{
+				Selector: fields.OneTermEqualSelector("spec.nodeName", node),
+			})
+	}
+	if err != nil {
+		return pods, fmt.Errorf("unable to fetch pods")
+	}
+	return pods, nil
+}
+
+func getClusterNames(cmd *cobra.Command, inputClusterName string, pods corev1.PodList, clusterLabel string) map[string]fdbv1beta2.None {
+	if inputClusterName != "" {
+		return map[string]fdbv1beta2.None{inputClusterName: {}}
+	}
+
+	clusterNames := make(map[string]fdbv1beta2.None)
+	for _, pod := range pods.Items {
+		clusterName, ok := pod.Labels[clusterLabel]
+		if !ok {
+			printStatement(cmd, fmt.Sprintf("could not fetch cluster name from Pod: %s\n", pod.Name), errorMessage)
```

Comment:
I don't think we want to reuse this method here, this will always add a ❌ in front of the error message.